src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs

```diff
@@ -176,69 +198,86 @@ internal static unsafe bool IsValidCanonical(char* name, int start, ref int end,
         // Return Invalid (-1) for failures.
         // If the address has less than three dots, only the rightmost section is assumed to contain the combined value for
         // the missing sections: 0xFF00FFFF == 0xFF.0x00.0xFF.0xFF == 0xFF.0xFFFF
-        internal static unsafe long ParseNonCanonical(char* name, int start, ref int end, bool notImplicitFile)
+        internal static unsafe long ParseNonCanonical<TChar>(TChar* name, int start, ref int end, bool notImplicitFile)
+            where TChar : unmanaged, IBinaryInteger<TChar>
         {
-            int numberBase = Decimal;
-            char ch;
-            long* parts = stackalloc long[4];
+            Debug.Assert(typeof(TChar) == typeof(char) || typeof(TChar) == typeof(byte));
+
+            int numberBase = IPv4AddressHelper.Decimal;
+            uint* parts = stackalloc uint[4];
             long currentValue = 0;
             bool atLeastOneChar = false;
 
             // Parse one dotted section at a time
             int dotCount = 0; // Limit 3
             int current = start;
+
             for (; current < end; current++)
             {
-                ch = name[current];
+                TChar ch = name[current];
+                int maxCharacterValue = '9';
                 currentValue = 0;
 
-                // Figure out what base this section is in
-                numberBase = Decimal;
-                if (ch == '0')
+                // Figure out what base this section is in, default to base 10.
+                // A number starting with zero should be interpreted in base 8 / octal
+                // If the number starts with 0x, it should be interpreted in base 16 / hex
+                numberBase = IPv4AddressHelper.Decimal;
+
+                if (ch == TChar.CreateTruncating('0'))
                 {
-                    numberBase = Octal;
```

Comment:
Previously this would have ended up with numberBase == Octal if current >= end. Now it'll end up as Decimal. I assume the stance is "that doesn't matter" because in both cases 0 means 0?

Reply:
Yes, that's right. This would be triggered by any IPv4 address ending in .0, so I'd personally expect that to be treated as a base-10 octet in its own right (rather than an octal prefix with a missing digit) - but in reality, it makes no difference. Zero is zero in all three bases we support.